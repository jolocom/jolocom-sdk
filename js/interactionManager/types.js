"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SigningType = exports.DecryptionType = exports.EncryptionType = exports.AuthorizationType = exports.EstablishChannelType = exports.FlowType = exports.InteractionRole = void 0;
var InteractionRole;
(function (InteractionRole) {
    InteractionRole["Requester"] = "requester";
    InteractionRole["Responder"] = "responder";
})(InteractionRole = exports.InteractionRole || (exports.InteractionRole = {}));
var FlowType;
(function (FlowType) {
    FlowType["Authentication"] = "Authentication";
    FlowType["CredentialShare"] = "CredentialShare";
    FlowType["CredentialOffer"] = "CredentialOffer";
    FlowType["Authorization"] = "Authorization";
    FlowType["EstablishChannel"] = "EstablishChannel";
    FlowType["Encrypt"] = "Encrypt";
    FlowType["Decrypt"] = "Decrypt";
    FlowType["Resolution"] = "Resolution";
    FlowType["Sign"] = "Sign";
})(FlowType = exports.FlowType || (exports.FlowType = {}));
var EstablishChannelType;
(function (EstablishChannelType) {
    EstablishChannelType["EstablishChannelRequest"] = "EstablishChannelRequest";
    EstablishChannelType["EstablishChannelResponse"] = "EstablishChannelResponse";
})(EstablishChannelType = exports.EstablishChannelType || (exports.EstablishChannelType = {}));
var AuthorizationType;
(function (AuthorizationType) {
    AuthorizationType["AuthorizationRequest"] = "AuthorizationRequest";
    AuthorizationType["AuthorizationResponse"] = "AuthorizationResponse";
})(AuthorizationType = exports.AuthorizationType || (exports.AuthorizationType = {}));
var EncryptionType;
(function (EncryptionType) {
    EncryptionType["EncryptionRequest"] = "EncryptionRequest";
    EncryptionType["EncryptionResponse"] = "EncryptionResponse";
})(EncryptionType = exports.EncryptionType || (exports.EncryptionType = {}));
var DecryptionType;
(function (DecryptionType) {
    DecryptionType["DecryptionRequest"] = "DecryptionRequest";
    DecryptionType["DecryptionResponse"] = "DecryptionResponse";
})(DecryptionType = exports.DecryptionType || (exports.DecryptionType = {}));
var SigningType;
(function (SigningType) {
    SigningType["SigningRequest"] = "SigningRequest";
    SigningType["SigningResponse"] = "SigningResponse";
})(SigningType = exports.SigningType || (exports.SigningType = {}));
//# sourceMappingURL=types.js.map