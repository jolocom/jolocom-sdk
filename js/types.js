"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_DELETE_AGENT_OPTIONS = exports.DEFAULT_EXPORT_OPTIONS = exports.EXPORT_SCHEMA_VERSION = exports.ChannelTransportType = exports.InteractionTransportType = void 0;
/**
 * @category Transports
 */
var InteractionTransportType;
(function (InteractionTransportType) {
    InteractionTransportType["QR"] = "QR";
    InteractionTransportType["Deeplink"] = "Deeplink";
    InteractionTransportType["HTTP"] = "HTTP";
    InteractionTransportType["Bluetooth"] = "Bluetooth";
    InteractionTransportType["NFC"] = "NFC";
})(InteractionTransportType = exports.InteractionTransportType || (exports.InteractionTransportType = {}));
/**
 * @category Transports
 */
var ChannelTransportType;
(function (ChannelTransportType) {
    ChannelTransportType["WebSockets"] = "WebSockets";
})(ChannelTransportType = exports.ChannelTransportType || (exports.ChannelTransportType = {}));
/**
 * @category Export/Import
 */
exports.EXPORT_SCHEMA_VERSION = "1.0.0";
/**
 * @category Export/Import
 */
exports.DEFAULT_EXPORT_OPTIONS = {
    credentials: true,
    interactions: true
};
/**
 * @category Delete
 */
exports.DEFAULT_DELETE_AGENT_OPTIONS = {
    encryptedWallet: true,
    identity: true,
    credentials: true,
    interactions: true,
};
//# sourceMappingURL=types.js.map