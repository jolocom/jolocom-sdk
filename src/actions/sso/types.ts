import { PublicProfileClaimMetadata } from 'cred-types-jolocom-core/types'

/**
 * @dev Simply using all claims required by the public profile
 */

export type IssuerPublicProfileSummary = PublicProfileClaimMetadata['claimInterface']

/**
 * @dev An identity summary is composed of a DID + all public info (currently public profile)
 */

export interface IdentitySummary {
  did: string
  publicProfile?: IssuerPublicProfileSummary
}

export interface PaymentRequestSummary {
  callbackURL: string
  requester: IdentitySummary
  receiver: {
    did: string
    address: string
  }
  requestJWT: string
  amount: number
  description: string
}
