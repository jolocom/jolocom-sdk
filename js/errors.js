"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SDKError = exports.ErrorCode = void 0;
var ErrorCode;
(function (ErrorCode) {
    ErrorCode["Unknown"] = "Unknown";
    ErrorCode["SaveExternalCredentialFailed"] = "SaveExtCred";
    ErrorCode["SaveCredentialMetadataFailed"] = "SaveCredMetadata";
    ErrorCode["TransportNotSupported"] = "UnsuportedTransport";
    ErrorCode["TokenExpired"] = "TokenExpired";
    ErrorCode["InvalidSignature"] = "InvalidSignature";
    ErrorCode["WrongDID"] = "WrongDID";
    ErrorCode["WrongNonce"] = "WrongNonce";
    ErrorCode["WrongFlow"] = "WrongFlow";
    ErrorCode["AuthenticationRequestFailed"] = "AuthRequest";
    ErrorCode["AuthenticationResponseFailed"] = "AuthResponse";
    ErrorCode["CredentialOfferFailed"] = "CredOffer";
    ErrorCode["CredentialsReceiveFailed"] = "CredsReceive";
    ErrorCode["CredentialRequestFailed"] = "CredRequest";
    ErrorCode["CredentialResponseFailed"] = "CredResponse";
    ErrorCode["ParseJWTFailed"] = "ParseJWT";
    ErrorCode["InvalidToken"] = "InvalidToken";
    ErrorCode["NoSuchInteraction"] = "NoSuchInteraction";
    ErrorCode["NoEntropy"] = "NoEntropy";
    ErrorCode["NoPassword"] = "NoPassword";
    ErrorCode["NoKeyProvider"] = "NoKeyProvider";
    ErrorCode["NoWallet"] = "NoWallet";
    ErrorCode["DecryptionFailed"] = "DecryptionFailed";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
class SDKError extends Error {
    constructor(code, origError) {
        super(code);
        this.origError = origError;
    }
}
exports.SDKError = SDKError;
SDKError.codes = ErrorCode;
//# sourceMappingURL=errors.js.map