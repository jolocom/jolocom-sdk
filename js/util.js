"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDeleteAgentOptions = exports.getExportAgentOptions = exports.jsonpath = exports.generateSecureRandomBytes = exports.generateIdentitySummary = void 0;
const crypto_1 = require("crypto");
const types_1 = require("./types");
/**
 * Given an identity, returns an object satisfying the {@link IdentitySummary} interface.
 * @dev Currently used with the {@link IssuerCard} component
 * @note In case the identity does not contain a Public Profile credential,
 * the function will return a minimal default which can be rendered.
 * @param identity - Instance of identity to generate the summary for
 */
exports.generateIdentitySummary = (identity) => {
    const { publicProfile, did } = identity;
    if (!publicProfile) {
        return {
            did,
        };
    }
    const { id, ...parsedProfile } = publicProfile.claim;
    return {
        did,
        publicProfile: parsedProfile,
    };
};
async function generateSecureRandomBytes(length) {
    return new Promise((resolve, reject) => {
        crypto_1.randomBytes(length, (err, bytes) => {
            if (err)
                reject(err);
            else
                resolve(bytes);
        });
    });
}
exports.generateSecureRandomBytes = generateSecureRandomBytes;
/**
 * Simple implementation of jsonpath that only supports very direct addressing
 * This takes a path p and and obj and returns the nested value denoted by the
 * path
 *
 * Example:
 * let obj = { some: { properties: ['a', 'b'] } }
 * jsonpath('$.some.properties.0', obj) === 'a'
 * jsonpath("$['some']['properties'][1]", obj) === 'b'
 * jsonpath("$.some", obj) === obj.some
 *
 * @param p - a path into the object, as in the example
 * @param obj - an object or array
 */
exports.jsonpath = function simpleJsonPath(p, obj) {
    let trimmedP = p.trim();
    if (trimmedP[0] != '$')
        return;
    let frags;
    if (trimmedP[1] === '.') {
        // remove the beginning '$.' and split on '.'
        frags = trimmedP.substring(2).split('.');
    }
    else if (trimmedP[1] === '[') {
        // remove the beginning '$' and replace separating '][' with ','
        trimmedP = trimmedP.substring(1).replace('][', ',');
        // result should look like a json list ['some',2]
        frags = JSON.parse(trimmedP);
    }
    else {
        return;
    }
    // go through the object key path and reduce to desired value
    return frags.reduce((obj, k) => {
        if (!obj)
            return;
        if (Array.isArray(obj)) {
            try {
                return obj[parseInt(k)];
            }
            catch (_a) {
                return;
            }
        }
        else if (typeof obj === 'object') {
            return obj[k];
        }
    }, obj);
};
exports.getExportAgentOptions = (options) => {
    return {
        ...types_1.DEFAULT_EXPORT_OPTIONS,
        ...options
    };
};
exports.getDeleteAgentOptions = (options) => ({ ...types_1.DEFAULT_DELETE_AGENT_OPTIONS, ...options });
//# sourceMappingURL=util.js.map