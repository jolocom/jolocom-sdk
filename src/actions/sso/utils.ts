import { IdentitySummary, IssuerPublicProfileSummary } from './types'
import { Identity } from 'jolocom-lib/js/identity/identity'
import { CredentialOfferRequest } from 'jolocom-lib/js/interactionTokens/credentialOfferRequest'
import { all, compose, either, isEmpty, isNil, map } from 'ramda'

/**
 * Given an identity, returns an object satisfying the {@link IdentitySummary} interface.
 * @dev Currently used with the {@link IssuerCard} component
 * @note In case the identity does not contain a Public Profile credential,
 * the function will return a minimal default which can be rendered.
 * @param identity - Instance of identity to generate the summary for
 */

export const generateIdentitySummary = (
  identity: Identity,
): IdentitySummary => {
  const { publicProfile, did } = identity
  if (!publicProfile) {
    return {
      did,
    }
  }
  const { id, ...parsedProfile } = publicProfile.claim
  return {
    did,
    publicProfile: parsedProfile as IssuerPublicProfileSummary,
  }
}

export const areRequirementsEmpty = (
  interactionToken: CredentialOfferRequest,
) =>
  compose(
    all(either(isNil, isEmpty)),
    map(interactionToken.getRequestedInputForType.bind(interactionToken)),
  )(interactionToken.offeredTypes)
