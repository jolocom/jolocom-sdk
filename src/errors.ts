export enum ErrorCode {
  Unknown = 'Unknown',

  SaveExternalCredentialFailed = 'SaveExtCred',
  SaveCredentialMetadataFailed = 'SaveCredMetadata',
  TransportNotSupported = 'UnsuportedTransport',

  TokenExpired = 'TokenExpired',
  InvalidSignature = 'InvalidSignature',
  WrongDID = 'WrongDID',
  WrongNonce = 'WrongNonce',
  WrongFlow = 'WrongFlow',

  AuthenticationRequestFailed = 'AuthRequest',
  AuthenticationResponseFailed = 'AuthResponse',

  CredentialOfferFailed = 'CredOffer',
  CredentialsReceiveFailed = 'CredsReceive',
  CredentialRequestFailed = 'CredRequest',
  CredentialResponseFailed = 'CredResponse',
  ParseJWTFailed = 'ParseJWT',

  InvalidToken = 'InvalidToken',

  NoEntropy = 'NoEntropy',
  NoPassword = 'NoPassword',
  NoKeyProvider = 'NoKeyProvider',
  NoWallet = 'NoWallet',
  DecryptionFailed = 'DecryptionFailed',
}

export class SDKError extends Error {
  public static codes = ErrorCode
  public origError?: Error

  public constructor(code: ErrorCode, origError?: Error) {
    super(code)
    this.origError = origError
  }
}
