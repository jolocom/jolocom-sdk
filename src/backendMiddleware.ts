import { IdentityWallet } from 'jolocom-lib/js/identityWallet/identityWallet'
import { Storage } from 'src/lib/storage/storage'
import { KeyChain, KeyChainInterface } from 'src/lib/keychain'
import { ConnectionOptions } from 'typeorm/browser'
import {
  createJolocomRegistry,
  JolocomRegistry,
} from 'jolocom-lib/js/registries/jolocomRegistry'
import { IpfsCustomConnector } from 'src/lib/ipfs'
import { jolocomContractsAdapter } from 'jolocom-lib/js/contracts/contractsAdapter'
import { jolocomEthereumResolver } from 'jolocom-lib/js/ethereum/ethereum'
import { jolocomContractsGateway } from 'jolocom-lib/js/contracts/contractsGateway'
import { JolocomLib } from 'jolocom-lib'
import { publicKeyToDID } from 'jolocom-lib/js/utils/crypto'
import { Identity } from 'jolocom-lib/js/identity/identity'
import { SoftwareKeyProvider } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'
import { generateSecureRandomBytes } from 'src/lib/util'
import { InteractionManager } from './lib/interactionManager/interactionManager'
import { BackendError, BackendMiddlewareErrorCodes } from './lib/errors/types'

export class BackendMiddleware {
  private _identityWallet!: IdentityWallet
  private _keyProvider!: SoftwareKeyProvider

  public storageLib: Storage
  public keyChainLib: KeyChainInterface
  public registry: JolocomRegistry
  public interactionManager: InteractionManager

  public constructor(config: {
    fuelingEndpoint: string
    typeOrmConfig: ConnectionOptions
  }) {
    this.storageLib = new Storage(config.typeOrmConfig)
    this.keyChainLib = new KeyChain()
    this.registry = createJolocomRegistry({
      ipfsConnector: new IpfsCustomConnector({
        host: 'ipfs.jolocom.com',
        port: 443,
        protocol: 'https',
      }),
      ethereumConnector: jolocomEthereumResolver,
      contracts: {
        adapter: jolocomContractsAdapter,
        gateway: jolocomContractsGateway,
      },
    })
    this.interactionManager = new InteractionManager(this)
  }

  public async initStorage(): Promise<void> {
    await this.storageLib.initConnection()
  }

  public get identityWallet(): IdentityWallet {
    if (this._identityWallet) return this._identityWallet
    throw new BackendError(BackendMiddlewareErrorCodes.NoWallet)
  }

  public get keyProvider(): SoftwareKeyProvider {
    if (this._keyProvider) return this._keyProvider
    throw new BackendError(BackendMiddlewareErrorCodes.NoKeyProvider)
  }

  public async prepareIdentityWallet(): Promise<IdentityWallet> {
    if (this._identityWallet) return this._identityWallet

    const encryptedEntropy = await this.storageLib.get.encryptedSeed()
    let encryptionPass
    try {
      encryptionPass = await this.keyChainLib.getPassword()
    } catch (e) {
      // This may fail if the application was uninstalled and reinstalled, as
      // the android keystore is cleared on uninstall, but the database may
      // still remain, due to having been auto backed up!
      // FIXME: Sentry.captureException(e)
    }

    if (encryptedEntropy && !encryptionPass) {
      // if we can't decrypt the encryptedEntropy, then reset the database
      console.warn('DROPPING OLD DB')
      await this.storageLib.resetDatabase()
    }

    if (!encryptedEntropy || !encryptionPass) {
      // If either encryptedEntropy or encryptionPass was missing, we throw
      // NoEntropy to signal that we cannot prepare an identityWallet instance due
      // to lack of a seed.
      // Note that the case of having an encryptionPass but no encryptedEntropy
      // is an uncommon edge case, but may potentially happen due to errors/bugs
      // etc
      throw new BackendError(BackendMiddlewareErrorCodes.NoEntropy)
    }

    this._keyProvider = new JolocomLib.KeyProvider(
      Buffer.from(encryptedEntropy, 'hex'),
    )
    const { jolocomIdentityKey: derivationPath } = JolocomLib.KeyTypes

    const userPubKey = this._keyProvider.getPublicKey({
      derivationPath,
      encryptionPass,
    })

    const didDocument = await this.storageLib.get.didDoc(
      publicKeyToDID(userPubKey),
    )

    if (didDocument) {
      const identity = Identity.fromDidDocument({ didDocument })

      // TODO Simplify constructor
      return (this._identityWallet = new IdentityWallet({
        identity,
        vaultedKeyProvider: this._keyProvider,
        publicKeyMetadata: {
          derivationPath,
          keyId: identity.publicKeySection[0].id,
        },
        contractsAdapter: this.registry.contractsAdapter,
        contractsGateway: this.registry.contractsGateway,
      }))
    } else {
      const identityWallet = await this.registry.authenticate(
        this._keyProvider,
        {
          encryptionPass,
          derivationPath,
        },
      )

      await this.storageLib.store.didDoc(identityWallet.didDocument)
      return (this._identityWallet = identityWallet)
    }
  }
  public async recoverIdentity(mnemonic: string): Promise<Identity> {
    const password = (await generateSecureRandomBytes(32)).toString('base64')
    this._keyProvider = JolocomLib.KeyProvider.recoverKeyPair(
      mnemonic,
      password,
    ) as SoftwareKeyProvider
    const { jolocomIdentityKey: derivationPath } = JolocomLib.KeyTypes

    const identityWallet = await this.registry.authenticate(this._keyProvider, {
      encryptionPass: password,
      derivationPath,
    })
    this._identityWallet = identityWallet
    await this.keyChainLib.savePassword(password)
    await this.storeIdentityData()
    return identityWallet.identity
  }

  public async createKeyProvider(encodedEntropy: string): Promise<void> {
    const password = (await generateSecureRandomBytes(32)).toString('base64')
    this._keyProvider = JolocomLib.KeyProvider.fromSeed(
      Buffer.from(encodedEntropy, 'hex'),
      password,
    )
    await this.keyChainLib.savePassword(password)
  }

  public async fuelKeyWithEther(): Promise<void> {
    const password = await this.keyChainLib.getPassword()
    await JolocomLib.util.fuelKeyWithEther(
      this.keyProvider.getPublicKey({
        encryptionPass: password,
        derivationPath: JolocomLib.KeyTypes.ethereumKey,
      }),
    )
  }

  public async createIdentity(): Promise<Identity> {
    const password = await this.keyChainLib.getPassword()
    this._identityWallet = await this.registry.create(
      this.keyProvider,
      password,
    )
    await this.storeIdentityData()
    return this._identityWallet.identity
  }

  private async storeIdentityData(): Promise<void> {
    const personaData = {
      did: this._identityWallet.identity.did,
      controllingKeyPath: JolocomLib.KeyTypes.jolocomIdentityKey,
    }
    await this.storageLib.store.persona(personaData)
    const encryptedSeedData = {
      // TODO: change to keyProvider.encryptedSeed when the library is updated
      // with a public getter for the encryptedSeed
      encryptedEntropy: this.keyProvider['encryptedSeed'].toString('hex'),
      timestamp: Date.now(),
    }
    await this.storageLib.store.encryptedSeed(encryptedSeedData)
    await this.storageLib.store.didDoc(this._identityWallet.didDocument)
  }
}
