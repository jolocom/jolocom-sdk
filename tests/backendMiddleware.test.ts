import * as util from 'src/lib/util'
import { BackendMiddleware } from 'src/backendMiddleware'
import { reveal, stub } from './utils'
import { ConnectionOptions } from 'typeorm/browser'
import data from 'tests/actions/registration/data/mockRegistrationData'
import { JolocomLib } from 'jolocom-lib'
import { IRegistry } from 'jolocom-lib/js/registries/types'
import { BackendError } from '../src/lib/errors/types'

const MockDate = require('mockdate')

const mockBackendMiddlewareConfig = {
  fuelingEndpoint: 'hptp://petrol1.station',
  typeOrmConfig: {} as ConnectionOptions,
}

describe('BackendMiddleware', () => {
  const { getPasswordResult, cipher, entropy, identityWallet } = data
  const keyChainLib = stub<BackendMiddleware['keyChainLib']>({
    getPassword: jest.fn().mockResolvedValue(getPasswordResult),
  })
  const decryptWithPass = jest.fn().mockReturnValue(entropy)
  const storageLib = {
    store: stub<BackendMiddleware['storageLib']['store']>(),
    get: stub<BackendMiddleware['storageLib']['get']>({
      encryptedSeed: jest.fn().mockResolvedValue(cipher),
    }),
    resetDatabase: jest.fn()
  }
  const registry = stub<IRegistry>()

  const createBackendMiddleware = (): BackendMiddleware => {
    const backendMiddleware = new BackendMiddleware(mockBackendMiddlewareConfig)
    Object.assign(backendMiddleware, {
      keyChainLib,
      encryptionLib: { decryptWithPass },
      storageLib,
      registry,
    })
    return backendMiddleware
  }

  jest
    .spyOn(util, 'generateSecureRandomBytes')
    .mockImplementation(() =>
      Promise.resolve(Buffer.from(getPasswordResult, 'base64')),
    )

  describe('prepareIdentityWallet', () => {
    const backendMiddleware = createBackendMiddleware()

    beforeEach(() => {
      stub.clearMocks(registry)
      stub.clearMocks(storageLib.get)
      stub.clearMocks(storageLib.store)
    })
    it('should throw NoEntropy if there is no encryptedEnntropy, regardless of encryptionPass', async () => {
      // no encryptedEntropy, no encryptionPass
      reveal(storageLib.get).encryptedSeed.mockResolvedValueOnce(null)
      reveal(keyChainLib).getPassword.mockRejectedValueOnce(new Error())
      let walletPromise = backendMiddleware.prepareIdentityWallet()
      await expect(walletPromise).rejects.toThrow(BackendError.codes.NoEntropy)

      // no encryptedEntropy, yes encryptionPass
      reveal(storageLib.get).encryptedSeed.mockResolvedValueOnce(null)
      walletPromise = backendMiddleware.prepareIdentityWallet()
      await expect(walletPromise).rejects.toThrow(BackendError.codes.NoEntropy)
      await expect(storageLib.resetDatabase).not.toHaveBeenCalled()
    })

    it('should clear database and throw NoEntropy if there is no encryptionPass but there is encryptedEntropy', async () => {
      reveal(keyChainLib).getPassword.mockRejectedValueOnce(new Error())
      const walletPromise = backendMiddleware.prepareIdentityWallet()
      return expect(walletPromise)
        .rejects.toThrow(BackendError.codes.NoEntropy)
        .then(() => {
          return expect(storageLib.resetDatabase).toHaveBeenCalled()
        })
    })

    it('should throw DecryptionFailed if decryption fails', async () => {
      decryptWithPass.mockReturnValueOnce(null)
      const walletPromise = backendMiddleware.prepareIdentityWallet()
      return expect(walletPromise).rejects.toThrowError()
    })

    it('should authenticate and cache the identity if not cached', async () => {
      reveal(storageLib.get).didDoc.mockResolvedValueOnce(null)
      reveal(registry).authenticate.mockResolvedValue(identityWallet)

      await backendMiddleware.prepareIdentityWallet()
      expect(reveal(registry).authenticate.mock.calls).toMatchSnapshot()
      expect(reveal(storageLib.get).didDoc.mock.calls).toMatchSnapshot()
      expect(storageLib.store.didDoc).toHaveBeenCalledWith(
        identityWallet.didDocument,
      )
      expect(() => backendMiddleware.identityWallet).not.toThrow()
    })

    it('should use cached identity if available', async () => {
      reveal(storageLib.get).didDoc.mockResolvedValueOnce(
        identityWallet.didDocument,
      )

      await backendMiddleware.prepareIdentityWallet()
      expect(reveal(registry).authenticate).not.toHaveBeenCalled()
      expect(storageLib.store.didDoc).not.toHaveBeenCalled()

      expect(() => backendMiddleware.identityWallet).not.toThrow()
      expect(backendMiddleware.identityWallet.identity.did).toMatch(
        identityWallet.didDocument.did,
      )
    })
  })

  describe('Identity Creation', () => {
    const backendMiddleware = createBackendMiddleware()
    let entropyData: { encryptedEntropy: string; timestamp: number }

    beforeAll(() => {
      MockDate.set(new Date(946681200000))
      entropyData = {
        encryptedEntropy: cipher,
        timestamp: Date.now(),
      }
    })

    afterAll(() => MockDate.reset())

    it('should throw NoKeyProvider from fuelKeyWithEther if no entropy was set', async () => {
      const fuelPromise = backendMiddleware.fuelKeyWithEther()
      return expect(fuelPromise).rejects.toThrow(
        BackendError.codes.NoKeyProvider,
      )
    })

    it('should throw NoKeyProvider from createIdentity if no entropy was set', async () => {
      const identityPromise = backendMiddleware.createIdentity()
      return expect(identityPromise).rejects.toThrow(
        BackendError.codes.NoKeyProvider,
      )
    })

    it('should throw NoKeyProvider from fuelKeyWithEther if no entropy was set', async () => {
      const fuelPromise = backendMiddleware.fuelKeyWithEther()
      return expect(fuelPromise).rejects.toThrow(
        BackendError.codes.NoKeyProvider,
      )
    })

    it('should createKeyProvider', async () => {
      await backendMiddleware.createKeyProvider(entropy)
      expect(() => backendMiddleware.keyProvider).not.toThrow()
      expect(
        backendMiddleware.keyProvider['encryptedSeed'].toString('hex'),
      ).toMatch(cipher)
    })

    it('should not store anything before the identity is registered', () => {
      expect(storageLib.store.didDoc).not.toHaveBeenCalled()
      expect(storageLib.store.persona).not.toHaveBeenCalled()
      expect(storageLib.store.encryptedSeed).not.toHaveBeenCalled()
    })

    it('should fuelKeyWithEther', async () => {
      const fuelSpy = jest
        .spyOn(JolocomLib.util, 'fuelKeyWithEther')
        .mockResolvedValueOnce(null)

      await backendMiddleware.fuelKeyWithEther()
      expect(fuelSpy.mock.calls).toMatchSnapshot()
    })

    it('should register the identity and store data on createIdentity', async () => {
      reveal(registry).create.mockResolvedValue(identityWallet)

      const identity = await backendMiddleware.createIdentity()

      expect(registry.create).toHaveBeenCalledWith(
        backendMiddleware.keyProvider,
        getPasswordResult,
      )
      expect(storageLib.store.didDoc).toHaveBeenCalledWith(
        identityWallet.didDocument,
      )
      expect(storageLib.store.persona).toHaveBeenCalledWith({
        did: identity.did,
        controllingKeyPath: JolocomLib.KeyTypes.jolocomIdentityKey,
      })
      expect(keyChainLib.savePassword).toHaveBeenCalledTimes(1)
      expect(keyChainLib.savePassword).toHaveBeenCalledWith(getPasswordResult)

      expect(storageLib.store.encryptedSeed).toHaveBeenCalledWith(entropyData)
    })
  })

  describe('Identity Recovery', () => {
    const backendMiddleware = createBackendMiddleware()
    let entropyData: { encryptedEntropy: string; timestamp: number }

    beforeAll(() => {
      MockDate.set(new Date(946681200000))
      entropyData = {
        encryptedEntropy: cipher,
        timestamp: Date.now(),
      }
      stub.clearMocks(keyChainLib)
    })

    afterAll(() => MockDate.reset())

    it('should recover the identity', async () => {
      reveal(registry).authenticate.mockResolvedValue(identityWallet)
      const identity = await backendMiddleware.recoverIdentity(
        'exhibit history avoid kit gaze pulse yellow portion hold lottery panda figure',
      )

      expect(registry.authenticate).toHaveBeenCalledWith(
        backendMiddleware.keyProvider,
        { derivationPath: "m/73'/0'/0'/0", encryptionPass: getPasswordResult },
      )

      expect(storageLib.store.didDoc).toHaveBeenCalledWith(
        identityWallet.didDocument,
      )
      expect(storageLib.store.persona).toHaveBeenCalledWith({
        did: identity.did,
        controllingKeyPath: JolocomLib.KeyTypes.jolocomIdentityKey,
      })
      expect(keyChainLib.savePassword).toHaveBeenCalledTimes(1)
      expect(keyChainLib.savePassword).toHaveBeenCalledWith(getPasswordResult)

      expect(storageLib.store.encryptedSeed).toHaveBeenCalledWith(entropyData)
    })
  })
})
