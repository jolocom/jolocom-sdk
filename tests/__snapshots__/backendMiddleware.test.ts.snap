// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackendMiddleware Identity Creation should fuelKeyWithEther 1`] = `
Array [
  Array [
    Object {
      "data": Array [
        3,
        224,
        237,
        157,
        101,
        36,
        234,
        73,
        97,
        64,
        32,
        79,
        228,
        11,
        209,
        94,
        213,
        7,
        251,
        36,
        13,
        220,
        129,
        82,
        186,
        68,
        39,
        118,
        160,
        126,
        38,
        233,
        185,
      ],
      "type": "Buffer",
    },
  ],
]
`;

exports[`BackendMiddleware prepareIdentityWallet should authenticate and cache the identity if not cached 1`] = `
Array [
  Array [
    SoftwareKeyProvider {
      "_encryptedSeed": Object {
        "data": Array [
          3,
          69,
          56,
          188,
          110,
          114,
          6,
          39,
          107,
          123,
          123,
          154,
          221,
          36,
          218,
          179,
          15,
          52,
          12,
          204,
          71,
          97,
          3,
          123,
          198,
          145,
          179,
          176,
          205,
          169,
          195,
          127,
        ],
        "type": "Buffer",
      },
      "_iv": Object {
        "data": Array [
          110,
          174,
          224,
          118,
          164,
          115,
          164,
          184,
          91,
          180,
          199,
          192,
          204,
          65,
          175,
          101,
        ],
        "type": "Buffer",
      },
    },
    Object {
      "derivationPath": "m/73'/0'/0'/0",
      "encryptionPass": "fuZJ4DVOSIi5fb+A6A9EDzNADsEWcHSu4jAiiWyYK+Q=",
    },
  ],
]
`;

exports[`BackendMiddleware prepareIdentityWallet should authenticate and cache the identity if not cached 2`] = `
Array [
  Array [
    "did:jolo:d129fdce5ff7d69878c001b0973d7e1a0ff2ae4cc7066d5a769e909c28c4149d",
  ],
]
`;
